### PR TITLE
cmd/evm: transaction validation tool

### LINKED ISSUE #23486: t8ntool / evm  transaction structure unit test interface
a protocol/command to run transaction structure unit tests. 
input:  transaction rlp, fork rule
output:  bool valid/invalid, recovered sender if any 

# Rationale

We have ttTransactionTests which are currently not maintained. the paradigm require a client opinion on whether provided transaction structure is valid or not. 

By asking a client if it thinks the provided transaction rlp is good or not (hypothetically) without any state, env info, we can revive the ttTransaction tests that test transaction structure parsing.

# Implementation

* Geth accept transaction rlp
* Checks the structure validity (number of fields for given transaction type, validity of RLP format)
* Checks that fields are in allowed range (0...2 ** 256), max data length, 2 ** 64 for gasLimit  (rules might be different depending on fork)
* Checks that gasLimit cover intrinsic transaction price (intrinsic transaction price + data bytes price)
* Recovers sender address

### PR DESCRIPTION
**Pull request by holiman**
This  is a tool to fix #23486, provide a tool for test-makers to validate transactions, based on a fork but no state. 

Follow-up to https://github.com/ethereum/go-ethereum/pull/23507

The input is a json string, the contents of which must be a hex-encoded RLP list. 
The elements in the RLP list can be anything -- a transaction or a `"hello world"`. If the input is not an RLP list, the input is considered invalid. 

For each element in the list, validation is performed as if it were a transaction, and each result is (for now) echoed on the output (subject to change -- it'll be json eventually). 

Example input data : 
```
[user@work evm]$ cat ./signed_txs.rlp
"0xf8d2b86702f864010180820fa08284d09411111111111111111111111111111111111111118080c001a0b7dfab36232379bb3d1497a4f91c1966b1f932eae3ade107bf5d723b9cb474e0a06261c359a10f2132f126d250485b90cf20f30340801244a08ef6142ab33d1904b86702f864010280820fa08284d09411111111111111111111111111111111111111118080c080a0d4ec563b6568cd42d998fc4134b36933c6568d01533b5adf08769270243c6c7fa072bf7c21eac6bbeae5143371eef26d5e279637f3bd73482b55979d76d935b1e9"
```

**Examples below no longer accurate, output format changed** 

On homestead: 
```
[user@work evm]$ ./evm t9n --input.txs ./signed_txs.rlp --state.fork=Homestead
error deriving sender, tx 0, transaction type not supported
error deriving sender, tx 1, transaction type not supported
```
On London:

```
[user@work evm]$ ./evm t9n --input.txs ./signed_txs.rlp --state.fork=London
sender tx 0: 0xd02d72E067e77158444ef2020Ff2d325f929B363
sender tx 1: 0xd02d72E067e77158444ef2020Ff2d325f929B363
```
cc @winsvega

### CODE REVIEW

**Comment by winsvega** [comment]
this can be the behaviour on t8n when pre and env are not provided.

**Comment by holiman** [comment]
> this can be the behaviour on t8n when pre and env are not provided.

I'm not sure. I think it may be better to have very well-specified one-job tools, instead of having implicit behaviour changes based on the presence/absence of inputs.

**Comment by winsvega** [comment]
I fhinished implementation on my side. 

One question. The transaction hash calculation would always stay the same as  "hash of (bytespayload)"
where bytes payload is rlp 0xf8...  for legacy  and  0x01f8....   0x02f8....   for typed transactions?

**Comment by holiman** [comment]
I think so yes. Want me to add hash to output aswell?

**Comment by winsvega** [comment]
the hash is just the hash of byte data that coming in?
`string const hash = "0x" + dev::toString(dev::sha3(fromHex(_rlp.asString())));`

so even if rlp is invalid it is still being calculated. 
I can calculate the hash. Just to be super safe for the future, (what if the hash logic would change) we can export the hash from t9n for hash function verification.

**Comment by holiman** [comment]
Yep, I agree

**Comment by holiman** [comment]
@winsvega hashes are now added. Please give it a spin and :+1: or :-1:

**Comment by winsvega** [comment]
doesn't compile

`cmd/evm/internal/t8ntool/transaction.go:79:34: undefined: fmt`

**Review by Ragnar00851 — APPROVED** [review]

**Comment by winsvega** [comment]
ok

**Comment by holiman** [comment]
@winsvega fixed, PTAL

**Comment by winsvega** [comment]
Yes, it's good

File: cmd/evm/internal/t8ntool/transaction.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	// We now have the transactions in 'body', which is supposed to be an
+	// rlp list of transactions
+	it, err := rlp.NewListIterator([]byte(body))
+	if err != nil {
+		return err
+	}
+	i := -1
+	var results []result
+	for it.Next() {
+		i++
```
Comment (MariusVanDerWijden) [inline cmd/evm/internal/t8ntool/transaction.go:121]: What do you need `i` for here?

**Review by MariusVanDerWijden — APPROVED** [review]
SGTM

File: cmd/evm/internal/t8ntool/transaction.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 The go-ethereum Authors
```
Comment (rjl493456442) [inline cmd/evm/internal/t8ntool/transaction.go:1]: 2021 :P

File: cmd/evm/internal/t8ntool/transaction.go (hunk 1 above)
Comment (holiman) [reply cmd/evm/internal/t8ntool/transaction.go:121]: well...
because.. reasons, ok!

damn it you folks

Comment (holiman) [reply cmd/evm/internal/t8ntool/transaction.go:121]: seriously though, thanks for checking!

**Pull request merged**